### PR TITLE
News page

### LINKED ISSUE #17: Set up news page
In this issue, I set up the news page using react query and news Api

### PR DESCRIPTION
**Pull request by motuncoded**
In this change, I set up the news card 
Closes #17

### CODE REVIEW

File: pages/queryKeys.ts [hunk 1]
```diff
@@ -11,6 +11,10 @@ const QueryKeys: QueryKey = {
   FETCH_COMMENTS: "FETCH_COMMENTS",
   FETCH_USERS: "FETCH_USERS",
   FETCH_NEWS: "FETCH_NEWS",
+
+
+
+
```
Comment (Asylcreek) [inline pages/queryKeys.ts]: Why?

File: pages/comments.tsx [hunk 2]
```diff
@@ -1,6 +1,5 @@
 import { inter } from "../styles/fonts";
-import { queryKeys } from "./queryKeys";
-import { useFetchComments } from "./hooks/useFetchComments"; // Adjust the path as necessary
+import useFetchComments from "./hooks/useFetchComments"; // Adjust the path as necessary
```
Comment (Asylcreek) [inline pages/comments.tsx]:
```suggestion
import useFetchComments from "./hooks/useFetchComments"; 
```

File: pages/hooks/useFetchComments.ts [hunk 3]
```diff
@@ -1,18 +1,16 @@
 import { useQuery } from "@tanstack/react-query";
-import { queryKeys } from "../queryKeys"; // Adjust the import path as necessary
+import { QueryKeys } from "../queryKeys"; // Adjust the import path as necessary
```
Comment (Asylcreek) [inline pages/hooks/useFetchComments.ts]:
```suggestion
import { QueryKeys } from "../queryKeys";
```

File: pages/hooks/useFetchComments.ts [hunk 4]
```diff
@@ -1,18 +1,16 @@
...
 
 const getComments = async () => {
   const COMMENTS_URL = "https://jsonplaceholder.typicode.com/comments";
```
Comment (Asylcreek) [inline pages/hooks/useFetchComments.ts]: This no longer needs to be in a variable

File: pages/hooks/useFetchNews.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { useQuery } from "@tanstack/react-query";
+import { QueryKeys } from "../queryKeys";
+
+const NEWS_API_KEY = "9bb1ce4f63d74db08d7433bd4bfd17a5";
+// Adjust the import path as necessary
```
Comment (Asylcreek) [inline pages/hooks/useFetchNews.ts]: Delete this line

File: pages/hooks/useFetchNews.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+const NEWS_URL = `https://newsapi.org/v2/top-headlines?country=us&apiKey=${NEWS_API_KEY}`;
+
+async function getNews(): Promise<any> {
+  try {
+    const response = await fetch(NEWS_URL);
+    return response.json();
+  } catch (error) {
+    console.error("Error fetching news:", error);
+    return { error: "Failed to fetch news" };
+  }
```
Comment (Asylcreek) [inline pages/hooks/useFetchNews.ts]: You should `try-catch` like this. ReactQuery would always return a success if you do. If you need to do stuff to the error, use the `onError` key that you can pass to `useQuery`

File: pages/hooks/useFetchUsers.ts [hunk 7]
```diff
@@ -1,21 +1,16 @@
 import { useQuery } from "@tanstack/react-query";
-import { queryKeys } from "../queryKeys";
-
-
+import { QueryKeys } from "../queryKeys";
 
 const fetchUsers = async () => {
-
   const USERS_URL = "https://jsonplaceholder.typicode.com/users";
```
Comment (Asylcreek) [inline pages/hooks/useFetchUsers.ts]: This needs not be in a variable as it is only being used inside this function

File: pages/news.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    return (
+      <div className="flex justify-center items-center">
+        <p>Error: {error?.message || "An unexpected error occurred."}</p>
+      </div>
+    );
+  }
+
+  return (
+    <div
+      className={`${poppins.className} flex justify-center items-center flex-col bg-[var(--main-bg-color-default)] text-[var(--main-color)] py-4`}
```
Comment (Asylcreek) [inline pages/news.tsx]: Shouldn't you load the `poppins` className in your layout file?

Also, setting the `poppins` className like this does not change the font-family of the element.

**Review by Asylcreek — CHANGES_REQUESTED** [review]

File: pages/_app.tsx [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import "@/styles/globals.css";
 import type { AppProps } from "next/app";
-import Navbar from "./components/Navbar";
 import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
+import Navbar from "./components/Navbar";
```
Comment (Asylcreek) [inline pages/_app.tsx]: This doesn't appear to be used

File: pages/comments.tsx [hunk 10]
```diff
@@ -1,14 +1,13 @@
...
+import useFetchComments from "./hooks/useFetchComments";
 
 type Comment = {
   id: number;
   body: string;
   name: string;
   email: string;
 };
-const Comments = () => {
+function CommentsPage() {
```
Comment (Asylcreek) [inline pages/comments.tsx]: Why did you change to using `function`?

File: pages/hooks/useFetchComments.ts (hunk 4 above)
Comment (Asylcreek) [reply pages/hooks/useFetchComments.ts]: You resolved this, but it actually isn't resolved yet

File: pages/hooks/useFetchNews.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+async function getNews(): Promise<any> {
+  try {
+    const response = await fetch(NEWS_URL);
+    if (!response.ok) {
+      throw new Error(`HTTP error! status: ${response.status}`);
+    }
+    return response.json();
+  } catch (error) {
+    throw error;
+  }
```
Comment (Asylcreek) [inline pages/hooks/useFetchNews.ts]: Throwing the error is equivalent to removing the `try-catch`

File: pages/hooks/useFetchUsers.ts [hunk 12]
```diff
@@ -1,21 +1,17 @@
 import { useQuery } from "@tanstack/react-query";
-import { queryKeys } from "../queryKeys";
-
+import { queryKeys } from "../query-keys";
 
+const USERS_URL = "https://jsonplaceholder.typicode.com/users";
```
Comment (Asylcreek) [inline pages/hooks/useFetchUsers.ts]: Why are you saving this in a variable?

File: pages/news.tsx [hunk 13]
```diff
@@ -0,0 +1,75 @@
+import React from "react";
```
Comment (Asylcreek) [inline pages/news.tsx]: You do not need to import React here

File: pages/posts.tsx [hunk 14]
```diff
@@ -1,3 +1,5 @@
+import React from "react";
```
Comment (Asylcreek) [inline pages/posts.tsx]: Why?

File: pages/posts.tsx [hunk 15]
```diff
@@ -1,3 +1,5 @@
...
+
 import { useState, useEffect, useRef } from "react";
 import { inter, roboto_serif } from "../styles/fonts";
```
Comment (Asylcreek) [inline pages/posts.tsx]: Shouldn't all your font initialisations be inside your root layout file?

Or are you saying that it is only this page that would make use of the `Inter` and `Roboto` fonts?

File: pages/query-keys.ts [hunk 16]
```diff
@@ -1,12 +1,19 @@
...
-
-
 //I think these should be all caps, the values. Because they are constants, and in JS world, we use CAPS_CAMEL_CASE to represent constants
 
-const queryKeys = {
+type QueryKey = {
+  FETCH_COMMENTS: string;
+  FETCH_USERS: string;
+  FETCH_NEWS: string;
+};
```
Comment (Asylcreek) [inline pages/query-keys.ts]: This type is not so necessary ... Typescript would infer the type of queryKeys as `Record<string, string>`

File: pages/users.tsx [hunk 17]
```diff
@@ -1,5 +1,6 @@
+import React from "react";
 import { inter } from "../styles/fonts";
```
Comment (Asylcreek) [inline pages/users.tsx]: This is the second place I am seeing this import, please check all your files that have font imports and move them into your root layout file.

PS. You can set up your next/fonts to use a CSS var, so you can just use that CSS var whenever you need to use that font.

File: pages/users.tsx [hunk 18]
```diff
@@ -18,11 +17,11 @@ type User = {
   email: string;
   username: string;
   phone: number;
-  address: Address; // Now the User type includes an address property.
+  address: Address;
 };
 
-const UsersList = () => {
-  const { data: users, isLoading, error } = useFetchUsers();
+function UsersPage() {
```
Comment (Asylcreek) [inline pages/users.tsx]: Why did you change to using `function`?

File: tsconfig.json [hunk 19]
```diff
@@ -16,6 +16,12 @@
       "@/*": ["./*"]
     }
   },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
+  "include": [
+    "next-env.d.ts",
+    "**/*.ts",
+    "**/*.tsx",
+    "next.config.js",
+    "pages/hooks/query-keys.ts"
```
Comment (Asylcreek) [inline tsconfig.json]: Nope. Remove this.

Also, you should only have files/folders inside the `pages` directory that you want to translate to actual routes. We had this conversation during the weekend.

`/pages/hooks/query-keys.ts` tells Next that it should prepare a route with `/hooks/query-keys`, which I am confident that you do not want.

**Review by Asylcreek — CHANGES_REQUESTED** [review]

File: pages/_app.tsx (hunk 9 above)
Comment (motuncoded) [reply pages/_app.tsx]: I have remove it form the index .tsx instead

File: pages/news.tsx [hunk 20]
```diff
@@ -1,5 +1,5 @@
 import Image from "next/image";
-import useFetchNews from "./hooks/useFetchNews";
+import useFetchNews from "../hooks/useFetchNews";
 import { poppins } from "../styles/fonts";
```
Comment (Asylcreek) [inline pages/news.tsx]: I am not expecting this here

Comment (Asylcreek) [reply pages/news.tsx]: You can (and should) create a layout component that imports all your fonts and attaches them to the DOM.

Not like this

**Review by Asylcreek — CHANGES_REQUESTED** [review]

**Review by Asylcreek — APPROVED** [review]

**Pull request merged**